zebra-state/src/service/check/anchors.rs

```diff
@@ -44,38 +44,56 @@ pub(crate) fn anchors_refer_to_earlier_treestates(
             // > either some earlier block’s final Sprout treestate, or to the interstitial
             // > output treestate of any prior JoinSplit description in the same transaction.
             //
             // https://zips.z.cash/protocol/protocol.pdf#joinsplit
-            let mut interstitial_roots: HashSet<sprout::tree::Root> = HashSet::new();
-
-            let mut interstitial_note_commitment_tree = parent_chain.sprout_note_commitment_tree();
-
-            for (n, joinsplit) in transaction.sprout_groth16_joinsplits().enumerate() {
-                for cm in joinsplit.commitments {
-                    interstitial_note_commitment_tree
-                        .append(cm)
-                        .expect("note commitment should be appendable to the tree");
-                }
-
-                interstitial_roots.insert(interstitial_note_commitment_tree.root());
+            //
+            // The FIRST JOINSPLIT in a transaction MUST refer to the output treestate
+            // of a previous block.
+            for (curr_joinsplit_pos, joinsplit) in joinsplits.iter().enumerate() {
+                let anchor = joinsplit.anchor;
 
                 tracing::debug!(?joinsplit.anchor, "observed sprout anchor");
 
-                if !parent_chain.sprout_anchors.contains(&joinsplit.anchor)
-                    && !finalized_state.contains_sprout_anchor(&joinsplit.anchor)
-                    && ((n == 0) || (!(n == 0) && !interstitial_roots.contains(&joinsplit.anchor)))
+                // Check if the Sprout treestate contains the anchor.
+                if !parent_chain.sprout_anchors.contains(&anchor)
+                    && !finalized_state.contains_sprout_anchor(&anchor)
                 {
-                    return Err(ValidateContextError::UnknownSproutAnchor {
-                        anchor: joinsplit.anchor,
-                    });
+                    // We incrementally build the interstitial treestate for
+                    // this transaction only if we didn't find the anchor in the
+                    // Sprout treestate.
+                    //
+                    // Note that when we have the FIRST JOINSPLIT, we skip the
+                    // whole `for` loop and return an error. This implicitly
+                    // covers the edge case for the FIRST JOINSPLIT in a
+                    // transaction.
+                    for prev_joinsplit in &joinsplits[prev_joinsplit_pos..curr_joinsplit_pos] {
+                        for cm in prev_joinsplit.commitments {
+                            interstitial_note_commitment_tree
+                                .append(cm)
+                                .expect("note commitment should be appendable to the tree");
+                        }
+
+                        interstitial_roots.insert(interstitial_note_commitment_tree.root());
+
+                        prev_joinsplit_pos = curr_joinsplit_pos;
+                    }
```

Comment:
This requires us to turn a O(nJoinSplits) into O(nJoinSplits!) algorithm, while saving us some possibly unnecessary Sprout note commitment tree updates (which only do SHA256Compress, vs Pedersen or Sinsemilla hash functions, which are much more expensive). 

IMO the original is (while doing extra work), clearer to understand the check we are trying to do: we are only ever   checking an anchor against previous-block-roots, or the set of interstitial tree roots up to this JoinSplit within this transaction.  If doing interstitial checks seems to be impacting sync, I would revisit improving this; right now, because of checkpointing and the anecdotal information about how/whether interstitial anchors were used in JoinSplits (especially post-Canopy), I think we can get away with the simple case.

Reply:
Oh, I noticed a big mistake. The way it is now is wrong and unintended. The assignment on line 92 should have been right under the loop, not inside it. Then it would have O(nJoinSplits). The advantage would be that the interstitial tree wouldn't be constructed at all if there weren't any JoinSplits that would need it. And if there were such JoinSplits, the tree would be constructed up to the last such one.

I agree it might be harder to read. Feel free to close this PR without merging though.

Reply:
OK. Closing, but keeping the branch around, just in case